Add external fluid property submodules

GiudGiud opened the pull request:
refs #27097

CIVET changes to get through:
- [x] https://github.com/idaholab/civet_recipes/pull/1739

Patches:
- [ ] https://github.inl.gov/ncrc/blue_crab/pull/272
- [ ] https://github.inl.gov/ncrc/dire_wolf/pull/293
- [x] https://github.inl.gov/ncrc/sockeye/pull/759
- [x] https://github.inl.gov/ncrc/pronghorn/pull/1266
- [x] https://github.inl.gov/ncrc/relap-7/pull/2569
- [ ] https://github.inl.gov/idaholab/sabertooth/pull/445

Apps submodules are kept around for the testing recipe to not fail. They are no longer used in compilation though.

TODO:
- [x] finish makefile
- [x] verify local testing is to code
OK: extensive unit testing. We could check coverage
- [x] verify documentation
Issues found:
  - [x] using the right folder structures for docs (wont show up at the right place)
  - only stubs!
- [x] create testing recipes
- [x] Fix non-unity build (missing headers)
- [x] get testing recipes through

Note:
- if you build the FP moose module, it depends on all contrib FPs
if you build a contrib FP, it wont build the other contrib FPs, but still depends on them if there are leftovers from the FP module compilation
This could complicate things for testing if moose is built only once

GiudGiud committed:
Add all fluid properties submodules
closes #27097

GiudGiud committed:
Add open FP submodules to Makefile
refs #27097

GiudGiud committed:
Add FP submodules to app file
Avoid infinite recursion
- we could also avoid registering multiple times but that s not breaking anything
refs #27097

GiudGiud committed:
Add FP submodule documentation and SQA doco to FP module

GiudGiud committed:
Update FP submodules past a round of fixes

GiudGiud committed:
Allow compilation when the FP modules are not checked out

GiudGiud committed:
Allow compilation from outside FP module
Makefile: only in FP module
fluid_properties.mk: both FP module and apps leveraging it

GiudGiud committed:
Use new FP submodules to attempt compilation

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/2175975/) on 65f3204 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/27027/site)

This comment will be updated on new commits.

GiudGiud committed:
Remove missing SQA config files
Remove others due to missing environment variables to find sqa_report from config file

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2176708/) on a838006 wanted to post the following:


## Framework coverage

Coverage did not change


## Modules coverage

### Fluid properties

<table>
  <thead>
    <tr>
      <th></th>
      <th>f4aebf</th>
      <th colspan=4>#27027 a83800</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.47%</td>
      <td>85.52%</td>
      <td>+0.05%</td>
      <td>85.71%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>7261</td>
      <td>7271</td>
      <td>+10</td>
      <td>6</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>1234</td>
      <td>1231</td>
      <td>-3</td>
      <td>1</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/27027/coverage_diff/fluid_properties)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/27027/coverage/fluid_properties)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/27027/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/27027/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/27027/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/27027/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/27027/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/27027/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/27027/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/27027/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/27027/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/27027/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/27027/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/27027/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/27027/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/27027/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/27027/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/27027/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/27027/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/27027/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/27027/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/27027/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/27027/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/27027/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/27027/coverage/scalar_transport)
- [`solid_mechanics`](https://mooseframework.inl.gov/docs/PRs/27027/coverage/solid_mechanics)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/27027/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/27027/coverage/stochastic_tools)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/27027/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/27027/coverage/xfem)

</details>

## Warnings

- `fluid_properties` new line coverage rate 85.71% is less than the suggested 90.0%



This comment will be updated on new commits.

GiudGiud committed:
Add more logic to find the FP submodules in the documentation build

GiudGiud committed:
Update submodules past SQA fix

GiudGiud committed:
Add FP submodules in main documentation:
- shows up in syntax list under their respective app
- shows up in source listing
refs #27097

GiudGiud commented:
Too big imo.

GiudGiud commented:
I m getting close here. Combined, FP, FP submodules build fine and so does doco. Working on the apps again now.

GiudGiud committed:
Fix makefile more
refs #27097

GiudGiud committed:
Set new FP submodules as external to prevent doco complete failure in downstream apps that would not download them
refs #27097

moosebuild commented:
Job [Conda build (Linux)](https://civet.inl.gov/job/2284484/) on 23bb5e0 : invalidated by @GiudGiud

zscalar hosing submodule downloads

GiudGiud committed:
Add empty categories for SQA as FP documents are being pulled in on accident
refs #27097

GiudGiud committed:
Add option to skip a module in SQA report
Add better error on failing json dump for sqa
Let SQA survive a missing path
refs #27097

GiudGiud committed:
Let moosedocs survive missing include directory
refs #27097

GiudGiud committed:
Update FP submodules to use more recent version with more SQA document
Add SQA categories for the FP submodules, because the sqa docs get picked up by the doco build
refs #27097

GiudGiud committed:
Update compilation variable names to not overlap with the one used by
RELAP7 Pronghorn etc, as the ordering of linking is different for them (at the moment)
refs #27097

GiudGiud commented:
This should be ready
@cticenhour to sign off on python doco/sqa changes
Note that a SQA doco build will require the submodules to be checked out.
Only two ways around it if we don't want that:
- change the yaml reader to not error on missing files
- dont include FP submodules in the config.yml sqa section, but then we need to find a way to exclude all the `.md` files from any doco build. Remove.yml isn't enough for some reason, I tried

A regular doco build does not require the submodules to be checked out

@joshuahansel for fluid properties related things

GiudGiud committed:
Remove SQA categories for FP submodule (we cannot pass No-Optional recipe otherwise)
Remove sqa files from doc content discovered by lexer
refs #27097

cticenhour commented:
In considering this in the last few minutes - if only the `No Optional` recipe is failing, I think we should remove SQA builds from that test. SQA is already tested via the `Documentation` recipe, and then once more at the master branch level. As far as SQA audit requirements, there isn't a scenario where `No Optional` makes sense for the relevant documentation - all is required to be there. I think we're safe to remove it from `No Optional` with justification.

GiudGiud commented:
so that's your preference?
Include the FPs in the SQA build and remove the sqa build from no-optional ?

cticenhour commented:
Yes - I prefer that we change the recipe, not the way we do SQA builds. Allowing optional items could lead to incorrect or incomplete SQA documentation. I can push up a PR to remove the arg from the recipe step.

GiudGiud commented:
I ll remove the last commit once the civet_recipe change is merged then. thanks!

GiudGiud commented:
Ready for review imo
let's get this in so I can send an angry email to the editors of wherever I sent the FP module paper for review

GiudGiud commented:
@cticenhour it turns out simply having the entry into the config.yml and not having the submodule checked out is enough to fail the doco build. Thoughts? Maybe leave the FPs outside of the SQA doco?
https://civet.inl.gov/pr/21959/

joshuahansel reviewed (commented):
Can/should we have tests in the MOOSE repo that use one or more of these modules, or are all of the tests only contained in those submodules' repos? I'm not sure what we do for the existing contrib submodules.

GiudGiud commented:
We can. I think for Saline we have one test. 
I'll add that this is a good idea to make sure they keep working (we only test that the executable builds and syntax is well registered from the moose side)

joshuahansel commented:
Ok, please add some very basic tests that use them. Otherwise, PR looks good to me.

GiudGiud committed:
Add testing for external submodules in fluid properties modules
Add a doco page for downloading the external submodule

refs #27097

GiudGiud commented:
@joshuahansel @cticenhour anything else you would like here?

joshuahansel commented on modules/fluid_properties/doc/content/source/fluidproperties/ExternalFluidPropertiesSubmodules.md:
```diff
@@ -0,0 +1,20 @@
+# External Fluid Properties Submodules
```
This page is good, but I don't see a link to it anywhere. Did I miss it? I would think that it should go on the fluid properties module page.

GiudGiud replied:
the page is available in the source and syntax dumps. We dont usually link to new objects specifically?

GiudGiud replied:
https://mooseframework.inl.gov/docs/PRs/27027/site/source/fluidproperties/ExternalFluidPropertiesSubmodules.html

joshuahansel replied:
It's in the source index, but I don't think it's in the syntax index, right? Imagine you're a user that is interested in getting air properties. You first browse the syntax index and you see an air fluid properties, so you click on that page, but it tells you nothing about how to get that class. Maybe after that you try looking at the fluid properties module page to see if it talks about it, but you don't find anything. You have no indication that what you need to be doing is going to the source index and looking for a page called "ExternalFluidPropertiesSubmodules".

GiudGiud replied:
Those other pages live on the submodules we can only link to those once moose has been updated to have this page first

GiudGiud replied:
By syntax page I just mean the object dump organized by syntax rather than by system/folder

joshuahansel replied:
> Those other pages live on the submodules we can only link to those once moose has been updated to have this page first

I'm confused - I'm able to go to https://mooseframework.inl.gov/docs/PRs/27027/site/fluidproperties/AirSBTLFluidProperties.html for example.

> By syntax page I just mean the object dump organized by syntax rather than by system/folder

I'm referring to https://mooseframework.inl.gov/docs/PRs/27027/site/syntax/index.html to be clear. I don't think the page would appear here without being located in `syntax/` and having a syntax.

GiudGiud replied:
> I'm confused - I'm able to go to

yes but it comes from the air git submodule. So to add a link there we would need to make a PR to the submodule not to moose

> I don't think the page would appear here without being located in syntax/ and having a syntax.

yeah no I would not want to do that. FluidProperties/ is still the syntax and we dont want to change that. 
By syntax I mean this page
https://mooseframework.inl.gov/docs/PRs/27027/site/syntax/FluidProperties/index.html

joshuahansel replied:
> yes but it comes from the air git submodule. So to add a link there we would need to make a PR to the submodule not to moose

I understand what you're saying now, thanks! So if the plan is to update all of the submodules with links to this page, then that's good. Best would be to additionally have a link on the module page - I can add that if you like.

joshuahansel commented:
Oh and I don't know if you like to do news entries now or later, but this one needs a news entry.

GiudGiud commented:
I like to do newsletter all at once at the end of the month

joshuahansel reviewed (approved):
Optional follow-up changes suggested.

GiudGiud commented:
Thinking of merging this next week after all the changes we got this week
